File: applications/rtktotalvariationdenoising/rtktotalvariationdenoising.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtktotalvariationdenoising_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"

#ifdef RTK_USE_CUDA
#  include "rtkCudaTotalVariationDenoisingBPDQImageFilter.h"
#else
#  include "rtkTotalVariationDenoisingBPDQImageFilter.h"
#endif

#include "rtkTotalVariationImageFilter.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtktotalvariationdenoising, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3; // Number of dimensions of the input image

#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
  using TVDenoisingFilterType = rtk::CudaTotalVariationDenoisingBPDQImageFilter;
#else
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
  using GradientOutputImageType = itk::Image<itk::CovariantVector<OutputPixelType, Dimension>, Dimension>;
  using TVDenoisingFilterType = rtk::TotalVariationDenoisingBPDQImageFilter<OutputImageType, GradientOutputImageType>;
#endif

  // Read input
  OutputImageType::Pointer input;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(input = itk::ReadImage<OutputImageType>(args_info.input_arg))

  // Compute total variation before denoising
  using TVFilterType = rtk::TotalVariationImageFilter<OutputImageType>;
  auto tv = TVFilterType::New();
  tv->SetInput(input);
  if (args_info.verbose_flag)
  {
    tv->Update();
    std::cout << "TV before denoising = " << tv->GetTotalVariation() << std::endl;
  }

  // Apply total variation denoising
  auto tvdenoising = TVDenoisingFilterType::New();
  tvdenoising->SetInput(input);
  tvdenoising->SetGamma(args_info.gamma_arg);
  tvdenoising->SetNumberOfIterations(args_info.niter_arg);

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(tvdenoising->GetOutput(), args_info.output_arg))

  // Compute total variation after denoising
  if (args_info.verbose_flag)
  {
    tv->SetInput(tvdenoising->GetOutput());
    tv->Update();
    std::cout << "TV after denoising = " << tv->GetTotalVariation() << std::endl;
  }

  return EXIT_SUCCESS;
}

File: applications/rtktotalvariationdenoising/rtktotalvariationdenoising.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Performs total variation denoising along the specified dimensions of a 3D image."
    )

    parser.add_argument(
        "--input", "-i", help="Input file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--gamma",
        "-g",
        help="TV term's weighting parameter",
        type=float,
        default=1.0,
    )
    parser.add_argument(
        "--niter",
        "-n",
        help="Number of iterations",
        type=int,
        default=5,
    )

    return parser


def process(args_info: argparse.Namespace):

    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Read input
    input_img = itk.imread(args_info.input)

    # Compute total variation before denoising
    if args_info.verbose:
        tv = rtk.TotalVariationImageFilter[OutputImageType].New()
        tv.SetInput(input_img)
        tv.Update()
        print(f"TV before denoising = {tv.GetTotalVariation()}")

    if hasattr(itk, "CudaImage"):
        tvdenoise = rtk.CudaTotalVariationDenoisingBPDQImageFilter.New()
        tvdenoise.SetInput(itk.cuda_image_from_image(input_img))
    else:
        GradPixelType = itk.CovariantVector[OutputPixelType, Dimension]
        GradImageType = itk.Image[GradPixelType, Dimension]
        tvdenoise = rtk.TotalVariationDenoisingBPDQImageFilter[
            OutputImageType, GradImageType
        ].New()
        tvdenoise.SetInput(input_img)

    tvdenoise.SetGamma(args_info.gamma)
    tvdenoise.SetNumberOfIterations(args_info.niter)

    # Write
    if args_info.verbose:
        print(f"Writing output to {args_info.output}...")

    writer = itk.ImageFileWriter[OutputImageType].New()
    writer.SetFileName(args_info.output)
    writer.SetInput(tvdenoise.GetOutput())
    writer.Update()

    # Compute total variation after denoising
    if args_info.verbose:
        tv_after = rtk.TotalVariationImageFilter[OutputImageType].New()
        tv_after.SetInput(tvdenoise.GetOutput())
        tv_after.Update()
        print(f"TV after denoising = {tv_after.GetTotalVariation()}")


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
